src/UID2.Client/UID2Encryption.cs

```diff
@@ -178,22 +190,52 @@ private static DecryptionResponse DecryptV3(byte[] encryptedId, KeyContainer key
                 return new DecryptionResponse(DecryptionStatus.UserOptedOut, null, established, siteId, siteKey.SiteId, identityType, advertisingTokenVersion, privacyBits.IsClientSideGenerated);
             }
 
-            if (enableDomainNameCheck && !IsDomainNameAllowedForSite(privacyBits, siteId, domainName, keys))
+            if (!IsDomainNameAllowedForSite(clientType, privacyBits, siteId, domainName, keys))
             {
-                return new DecryptionResponse(DecryptionStatus.DomainNameCheckFailed, null, established, siteId, siteKey.SiteId, identityType, advertisingTokenVersion,
-                    privacyBits.IsClientSideGenerated);
+                return new DecryptionResponse(DecryptionStatus.DomainNameCheckFailed, null, established, siteId, siteKey.SiteId, identityType, advertisingTokenVersion, privacyBits.IsClientSideGenerated);
             }
 
+            if (!DoesTokenHaveValidLifetime(clientType, keys, established, expiry, now))
+                return new DecryptionResponse(DecryptionStatus.InvalidTokenLifetime, null, established, siteId, siteKey.SiteId, identityType, advertisingTokenVersion, privacyBits.IsClientSideGenerated);
+
             return new DecryptionResponse(DecryptionStatus.Success, idString, established, siteId, siteKey.SiteId, identityType, advertisingTokenVersion, privacyBits.IsClientSideGenerated);
         }
 
-        private static bool IsDomainNameAllowedForSite(PrivacyBits privacyBits, int siteId, string domainName, KeyContainer keys)
+        private static bool DoesTokenHaveValidLifetime(ClientType clientType, KeyContainer keys, DateTime established, DateTime expiry, DateTime now)
         {
-            if (!privacyBits.IsClientSideGenerated)
+            long maxLifetimeSeconds;
+            switch (clientType)
             {
-                return true;
+                case ClientType.Bidstream:
+                    maxLifetimeSeconds = keys.MaxBidstreamLifetimeSeconds;
+                    break;
+                case ClientType.Sharing:
+                    maxLifetimeSeconds = keys.MaxSharingLifetimeSeconds;
+                    break;
+                default: //Legacy
+                    return true;
             }
 
+            return DoesTokenHaveValidLifetimeImpl(established, expiry, now, maxLifetimeSeconds, keys.AllowClockSkewSeconds);
+        }
+
+
+        private static bool DoesTokenHaveValidLifetimeImpl(DateTime established, DateTime expiry, DateTime now, long maxLifetimeSeconds, long allowClockSkewSeconds)
+        {
+            if ((expiry - established).TotalSeconds > maxLifetimeSeconds)
+                return false;
+
+            return (established - now).TotalSeconds <= allowClockSkewSeconds;
```

Comment:
do we expect (established - now).TotalSeconds can be negative here?

Reply:
yes, if no clock skew then we expect it to be negative (token created in the past), which will correctly return true